source/MaterialXGenShader/ShaderNode.cpp

```diff
@@ -383,38 +414,78 @@ void ShaderNode::setPaths(const Node& node, const NodeDef& nodeDef, bool include
-                input->setValue(nodeValue->getResolvedValue());
-            }
+        // Early out if no metadata is registered.
+        return;
+    }
 
-            InputPtr inputElem = nodeValue->asA<Input>();
-            if (inputElem)
+    // Set metadata on the node according to the nodedef attributes.
+    ShaderMetadataVecPtr nodeMetadataStorage = getMetadata();
+    for (const string& nodedefAttr : nodeDef.getAttributeNames())
+    {
+        const ShaderMetadata* metadataEntry = registry->findMetadata(nodedefAttr);
+        if (metadataEntry)
+        {
+            const string& attrValue = nodeDef.getAttribute(nodedefAttr);
+            if (!attrValue.empty())
             {
-                input->setChannels(inputElem->getChannels());
+                ValuePtr value = Value::createValueFromStrings(attrValue, metadataEntry->type->getName());
+                if (!value)
+                {
+                    value = metadataEntry->value;
+                }
+                if (value)
+                {
+                    if (!nodeMetadataStorage)
+                    {
+                        nodeMetadataStorage = std::make_shared<ShaderMetadataVec>();
+                        setMetadata(nodeMetadataStorage);
+                    }
+                    nodeMetadataStorage->push_back(ShaderMetadata(metadataEntry->name, metadataEntry->type, value));
+                }
             }
         }
     }
 
-    // Set implementation specific values.
-    if (_impl)
+    // Set metadata on inputs according to attributes on the nodedef's inputs
+    for (const ValueElementPtr& nodedefPort : nodeDef.getActiveValueElements())
     {
-        _impl->setValues(node, *this, context);
+        ShaderInput* input = getInput(nodedefPort->getName());
+        if (input)
+        {
+            ShaderMetadataVecPtr inputMetadataStorage = input->getMetadata();
+
+            for (const string& nodedefPortAttr : nodedefPort->getAttributeNames())
+            {
+                const ShaderMetadata* metadataEntry = registry->findMetadata(nodedefPortAttr);
+                if (metadataEntry)
+                {
+                    const string& attrValue = nodedefPort->getAttribute(nodedefPortAttr);
+                    if (!attrValue.empty())
+                    {
+                        const TypeDesc* type = metadataEntry->type ? metadataEntry->type : input->getType();
+                        ValuePtr value = Value::createValueFromStrings(attrValue, type->getName());
+                        if (!value)
+                        {
+                            value = metadataEntry->value;
+                        }
+                        if (value)
+                        {
+                            if (!inputMetadataStorage)
+                            {
+                                inputMetadataStorage = std::make_shared<ShaderMetadataVec>();
+                                input->setMetadata(inputMetadataStorage);
```

Comment:
Am I correct in assuming that the nodedef or the node can have have meta data defined and the node data "wins" -- like other logic which follows this pattern.   If so:
1. Not here but in "validation" does there need to be checking to see that the metadata between nodedef and node matches.
2. Would it be better to set according to the node first and skip the nodedef meta data if already set here?

Reply:
I only added support for exporting metadata set on the nodedef, because I didn't see a use-case for overriding this on the node instance. If we find a use-case for that maybe we could add this later.